### PR TITLE
Windows: Set service recovery options

### LINKED ISSUE #27696: dockerd.exe --register-service should set a service restart action
**Description**

`dockerd.exe --register-service` is used to create the "Docker" service and set it to automatically start when a Windows system boots up. This was done as an alternative to using something like NSSM and uses the service control manager built into Windows. 

One key feature is that it can restart services automatically, but its not configured by default. My ask is that we enable that by default to:
- Restart the service after 60 seconds on the first or second failure
- Restart the failure counter after a day

Infinite restarts can spend too much CPU cycles in the case of a crash, so I don't think that should be the default.

**Steps to reproduce the issue:**
1. `dockerd.exe --register-service`
2. `net start Docker`
3. `get-process -name dockerd | stop-process -Force`
4. Wait ~1 minute
5. `docker info`
6. `Get-Service Docker`

**Describe the results you received:**

`docker info` fails:
```none
error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/info: open //./pipe/docker_engine: The system cannot find the file specified.```

```none
 Get-Service Docker

Status   Name               DisplayName
------   ----               -----------
Stopped  Docker             Docker Engine
```

**Describe the results you expected:**
`Get-Service Docker` should show running

`docker info` shouldn't give an error

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 17
Server Version: 1.12.2-cs2-ws-beta
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: nat null overlay
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.321.amd64fre.rs1_release_inmarket.161004-2338)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 7 GiB
Name: plang-ws16-1
ID: 6Z4K:IBTG:Q3H3:YGUR:63I6:A6CT:GPF5:G3CN:COTV:63Q2:F7G5:ACBO
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
No dependency - this repros on any Windows machine, physical or virtual

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@jstarks @patricklang This sets the service restart options for the Windows daemon running as a service. Fixes #27696. 

@friism FYI - one for cherry-picking.

### CODE REVIEW

File: cmd/dockerd/service_windows.go [hunk 1]
```diff
@@ -183,6 +184,37 @@ func registerService() error {
...
+		Type  uint32
+		Delay uint32
+	}
+	t := []SC_ACTION{
+		{Type: SC_ACTION_RESTART, Delay: uint32(60000)},
+		{Type: SC_ACTION_RESTART, Delay: uint32(60000)},
+		{Type: SC_ACTION_NONE},
+	}
+	lpInfo := SERVICE_FAILURE_ACTIONS{ResetPeriod: uint32(60 * 60 * 24), ActionsCount: uint32(3), Actions: uintptr(unsafe.Pointer(&t[0]))}
+	windows.ChangeServiceConfig2(s.Handle, SERVICE_CONFIG_FAILURE_ACTIONS, (*byte)(unsafe.Pointer(&lpInfo)))
```
Comment (jstarks) [inline cmd/dockerd/service_windows.go]: Check error code?

**Comment by friism** [comment]
Please consider how this will influence the installer script: https://github.com/OneGet/MicrosoftDockerProvider/blob/developer/DockerMsftProvider.psm1#L312

File: cmd/dockerd/service_windows.go [hunk 2]
```diff
@@ -183,6 +184,37 @@ func registerService() error {
...
+		ActionsCount uint32
+		Actions      uintptr
+	}
+
+	type SC_ACTION struct {
+		Type  uint32
+		Delay uint32
+	}
+	t := []SC_ACTION{
+		{Type: SC_ACTION_RESTART, Delay: uint32(60000)},
```
Comment (jstarks) [inline cmd/dockerd/service_windows.go]: For clarity, probably better to write these durations as something like:

60 * time.Second / time.Millisecond

**Comment by jhowardmsft** [comment]
@friism Indeed - that's what @patricklang is trying to ratify. He asked for this change 😇

File: cmd/dockerd/service_windows.go (hunk 2 above)
Comment (jhowardmsft) [inline cmd/dockerd/service_windows.go]: Updated

File: cmd/dockerd/service_windows.go [hunk 3]
```diff
@@ -183,6 +185,40 @@ func registerService() error {
...
+	type scAction struct {
+		Type  uint32
+		Delay uint32
+	}
+	t := []scAction{
+		{Type: scActionRestart, Delay: uint32(60 * time.Second / time.Millisecond)},
+		{Type: scActionRestart, Delay: uint32(60 * time.Second / time.Millisecond)},
+		{Type: scActionNone},
+	}
+	lpInfo := serviceFailureActions{ResetPeriod: uint32(24 * 60 * 60), ActionsCount: uint32(3), Actions: uintptr(unsafe.Pointer(&t[0]))}
```
Comment (jhowardmsft) [inline cmd/dockerd/service_windows.go]: Updated (note `/time.Second`, not `/time.Millisecond` is the correct unit conversion factor here)

**Comment by jstarks** [comment]
LGTM

**Comment by johnstep** [comment]
LGTM

**Comment by jhowardmsft** [comment]
@thaJeztah @vdemeester PTAL. 2 LGTMs but from non-maintainers.

**Review by vdemeester — APPROVED** [review]
LGTM 👼

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**